prog/heatmap.go

```diff
@@ -16,36 +16,41 @@ import (
 //  2. Select random indices according to the probability distribution:
 //     `idx := hm.ChooseLocation(r)`.
 type Heatmap interface {
-	ChooseLocation(r *rand.Rand) int
-	Size() int
+	NumMutations() int
+	ChooseLocation() int
 }
 
 // Generic heatmaps model a probability distribution based on sparse data,
 // prioritising selection of regions which are not a single repeated byte. It
 // views data as a series of chunks of length `granularity`, ignoring chunks
 // which are a single repeated byte. Indices are chosen uniformly amongst the
 // remaining "interesting" segments.
-func MakeGenericHeatmap(data []byte) Heatmap {
+func MakeGenericHeatmap(data []byte, r *rand.Rand) Heatmap {
 	if len(data) == 0 {
 		panic("cannot create a GenericHeatmap with no data")
 	}
-	var hm GenericHeatmap
+	hm := &GenericHeatmap{
+		r: r,
+	}
 	hm.length, hm.segments = calculateLengthAndSegments(data, granularity)
-	return &hm
+	return hm
+}
+
+func (hm *GenericHeatmap) NumMutations() int {
+	// At least one mutation + up to about one mutation every 4 KB of heatmap size
+	// + up to 4 mutations at random so that even small images can get more than one.
+	return 1 + hm.r.Intn(hm.length/(4<<10)+1) + hm.r.Intn(5)
```

Comment:
Maybe add some cap on the value? If we, for some reason, get a very populated image, we can end up doing tons of mutations each time.

Reply:
Good point. Done. PTAL